federation/schema.go

```diff
@@ -76,6 +76,43 @@ func validateFederationKeys(serviceNames []string, serviceSchemasByName map[stri
 	return nil
 }
 
+// validateFieldsReturningFederatedObject checks that if an object is returned by a field func it can not be a shadow object type
+func validateFieldsReturningFederatedObject(serviceNames []string, serviceSchemasByName map[string]*IntrospectionQueryResult, types map[string]graphql.Type, fieldInfos map[*graphql.Field]*FieldInfo) error {
+	for _, service := range serviceNames {
+		for _, typ := range serviceSchemasByName[service].Schema.Types {
+			if typ.Kind == "OBJECT" {
+				for _, field := range typ.Fields {
+					// Check that the field's return type is an object
+					fieldReturnType := getRootType(field.Type)
+					if fieldReturnType.Kind == "OBJECT" {
+						// Error if it is a shadow object. To check this
+						// (1) Look through all the fields on the object to see if there is a federation field (__federation) and that
+						// the federation field is not on the current service
+						// (2) Look through all the fields on the federation object to see if it has a field for <ObjectType>-<Service>
+						returnObj, _ := types[fieldReturnType.Name].(*graphql.Object)
```

Comment:
we can remove the  `_` if we are sure the `returnObj` must be a `graphql.Object`.